scylla.go

```diff
@@ -484,53 +488,128 @@ func (p *scyllaConnPicker) GetCustomDialer() Dialer {
+	// The returned connection must be directly usable for CQL protocol,
+	// specifically DialShard is responsible also for setting up the TLS session if needed.
+	DialShard(ctx context.Context, host *HostInfo, shardID, nrShards int) (*DialedHost, error)
 }
 
 // A dialer which dials a particular shard
-type scyllaOneShardDialer struct {
-	shardAwareAddress string
-	shardID           int
-	nrShards          int
-	dialer            Dialer
+type scyllaDialer struct {
+	dialer    Dialer
+	logger    StdLogger
+	tlsConfig *tls.Config
+	cfg       *ClusterConfig
 }
 
 const scyllaShardAwarePortFallbackDuration time.Duration = 5 * time.Minute
 
-func (sosd *scyllaOneShardDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
-	iter := newScyllaPortIterator(sosd.shardID, sosd.nrShards)
+func (sd *scyllaDialer) DialHost(ctx context.Context, host *HostInfo) (*DialedHost, error) {
+	ip := host.ConnectAddress()
+	port := host.Port()
+
+	if !validIpAddr(ip) {
+		return nil, fmt.Errorf("host missing connect ip address: %v", ip)
+	} else if port == 0 {
+		return nil, fmt.Errorf("host missing port: %v", port)
+	}
+
+	addr := host.HostnameAndPort()
+	conn, err := sd.dialer.DialContext(ctx, "tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return sd.wrapTLS(conn, addr)
+}
+
+func (sd *scyllaDialer) DialShard(ctx context.Context, host *HostInfo, shardID, nrShards int) (*DialedHost, error) {
+	ip := host.ConnectAddress()
+	port := host.Port()
+
+	if !validIpAddr(ip) {
+		return nil, fmt.Errorf("host missing connect ip address: %v", ip)
+	} else if port == 0 {
+		return nil, fmt.Errorf("host missing port: %v", port)
+	}
+
+	iter := newScyllaPortIterator(shardID, nrShards)
+
+	addr := host.HostnameAndPort()
+
+	var shardAwarePort uint16
+	if sd.tlsConfig != nil {
+		shardAwarePort = host.ScyllaShardAwarePortTLS()
+	} else {
+		shardAwarePort = host.ScyllaShardAwarePort()
+	}
+
+	var shardAwareAddress string
+	if shardAwarePort != 0 {
+		tIP, tPort := sd.cfg.translateAddressPort(host.UntranslatedConnectAddress(), int(shardAwarePort))
+		shardAwareAddress = net.JoinHostPort(tIP.String(), strconv.Itoa(tPort))
+	}
 
 	if gocqlDebug {
-		Logger.Printf("scylla: connecting to shard %d", sosd.shardID)
+		sd.logger.Printf("scylla: connecting to shard %d", shardID)
+	}
+
+	conn, err := sd.dialShardAware(ctx, addr, shardAwareAddress, iter)
+	if err != nil {
+		return nil, err
+	}
+
+	return sd.wrapTLS(conn, addr)
+}
+
+func (sd *scyllaDialer) wrapTLS(conn net.Conn, addr string) (*DialedHost, error) {
+	if sd.tlsConfig != nil {
+		tlsConfig := tlsConfigForAddr(sd.tlsConfig, addr)
+		tconn := tls.Client(conn, tlsConfig)
+		if err := tconn.Handshake(); err != nil {
+			conn.Close()
+			return nil, err
+		}
+		conn = tconn
 	}
 
+	return &DialedHost{
+		Conn:            conn,
+		DisableCoalesce: sd.tlsConfig != nil, // write coalescing doesn't work with framing on top of TCP like in TLS.
```

Comment:
I do not understand what you mean here. Maybe it's a matter of the implementation but what exactly does not work?

Reply:
I copied this comment from the original code:

https://github.com/gocql/gocql/blob/89e6637b0c0191cb61ba4b9c1556d821ca55a8c7/connectionpool.go#L119

git blame points to the following explanation: https://github.com/gocql/gocql/commit/93aced8f230cbe935a771cf0918959b82b25003e

TLDR: TLS connnections can't use writev, so don't reduce the number of syscalls.

Reply:
I have updated the comment here and also for the HostDialer interface.